src/buffer/out/textBufferCellIterator.cpp

```diff
@@ -95,18 +95,66 @@ bool TextBufferCellIterator::operator!=(const TextBufferCellIterator& it) const
 TextBufferCellIterator& TextBufferCellIterator::operator+=(const ptrdiff_t& movement)
 {
     ptrdiff_t move = movement;
-    auto newPos = _pos;
-    while (move > 0 && !_exceeded)
+    if (move < 0)

```

Comment:
Let's say we are doing the "1 million yes" benchmark with a typical 120x80 windows. The `+=` operator will execute 120 x 1000000 = 120000000 times. In this particular method, I think it's safe to say that *every single instruction counts*. 

First, this line here is the early branching to avoid the very rare case (move < 0) and leave it to `-=` operator. Next I replace the `IncrementInBounds` with an simplified,  cache-friendly version, without compromising functionally. And finally I replace the `_GenerateView` with fine-grained `Update` methods.

Reply:
if `_GenerateView` is that bad... I feel like we should replace all usages of it, not just this one `+=` usage of it.

Same goes for `IncrementInBounds` and `DecrementInbounds`.

I'm OK with it being more complex inside this operator specifically as long as it's well documented why things are ordered and computed in the way they are (like specifically calling out which items below are taking advantage of cache improvements so no one tries to refactor it out to be "easier to read" but significantly worse performance in the future.)

Reply:
Besides the "early optimization is the root of all evil" cliche, I'd like to point out that `_GenerateView`, `IncrementInBounds` and `DecrementInbounds` are not actually expensive methods in terms of CPU costs. 

Take a look at my screenshot and you will see `WalkInBoundsCircular` (which is essentially what `IncrementInBounds ` calls) is only 1.35% of the CPU time. And it's in fact accurate, because that method itself isn't guilty for the performance drop. It's the unfortune *combination* of a whole lot of other methods in the hot paths that ruins the icache (instruction cache) at a particular point. 

Another example: I've actually tried to remove the code inside this method step by step. At some point, even a single line like `_pos = newPos` can show up >7% in the trace, which is obviously not a really expensive operation, but triggers the cache miss.

In conclusion, it takes accurate benchmarking  to "port" this optimization somewhere else. And the optimization may very well be NOT necessary in other code paths.

I got the inspiration from @lhecker  and he might want to add his opinions here, too. I'll try to add more comments as documentation to prevent "easier to read" refactoring in the future.

Reply:
Great justification. I'm fine with you not optimizing the other ones. Just call out some of this in the comments in the code and I'm sold.